### DISCUSSION TITLE
Concatination of python error messages.

### DISCUSSION THREAD

**Discussion by synapticonashaikh** [micropython/micropython#11954] [General]
Hi all, I'm working on a project where I'm redirecting all the prints and error logs into a log file. And to do so, I modified the function `mp_hal_stdout_tx_strn`. And in my log file, I store each message on a new line (for better understanding and visibility). The code worked fine with the `print` function, but I noticed some unusual behaviour when I tried to do the same thing with error messages. I debugged a little and saw that when an error occurs, the executor calls this function `mp_hal_stdout_tx_strn` several times (`8 to 9 times`). This results in broken logs in my log file. So, could anyone please tell me if there is a way or method to collectively concatenate the error messages? And then, I can push it into the log file.

```
Traceback (most recent call last):
(message 1)  File "(message 2)<stdin>(message 3)"(message 4)
(message 5)IndentationError(message 6): (message 7)unexpected indent(message 8)
(message 9)

```
```
/*updated function*/
FILENAME = "logging_curr.log"

void mp_hal_stdout_tx_strn(const char *str, mp_uint_t len) 
{ 
unsigned char d;
  //if(len != RESET )
  //d = write(STDOUT_FILENO, str, len); fflush(stdout);
   
  if(len != RESET )
    d = write_in_logs(FILENAME, str, len);  
}
```
![image](https://github.com/micropython/micropython/assets/115145853/85942a78-4707-4f20-aea3-2fa6315013dd)

**Comment by kjm1102** [comment]
You can log exceptions to a file in upython eg
```
def _logerr(er):
  try:
    from uio import StringIO; import sys; s=StringIO(); sys.print_exception(er, s)
    s=s.getvalue(); L=s.split('\n'); num=len(L)
    cause=L[num-3].split(','); line=cause[1]; err=L[num-2]; txt=line+' '
  except Exception as e: sys.print_exception(e); err=s; txt='_logerr prob '
  msg=err.replace('"',"").replace("'",''); f=open('errlog','a'); f.write(txt+msg+'\n'); f.close()
  with open('errlog', 'r') as f: print(f.read())
  
try:
  1/0
except Exception as e: _logerr(e)
```

**Comment by bidrohini** [comment]
I can see there is a mention of an indentation error in your code. Please check in which line it occurred and correct.

**Reply by synapticonashaikh** [reply]
Sir, you didn't get the question at all. I added the error in the script to check the logging process or output.
